lib/view_component/global_output_buffer.rb

```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
+
+module ViewComponent
+  module GlobalOutputBuffer
+    def render_in(view_context, &block)
+      unless view_context.output_buffer.is_a?(OutputBufferStack)
+        # use instance_variable_set here to avoid triggering the code in the #output_buffer= method below
+        view_context.instance_variable_set(:@output_buffer, OutputBufferStack.new(view_context.output_buffer))
+      end
+
+      @output_buffer = view_context.output_buffer
+
+      super(view_context, true, &block)
+    end
+
+    def perform_render
+      # HAML unhelpfully assigns to @output_buffer directly, so we hold onto a reference to
+      # it and restore @output_buffer when the HAML engine is finished. In non-HAML cases,
+      # @output_buffer and orig_buf will point to the same object, making the reassignment
+      # statements no-ops.
+      orig_buf = @output_buffer
+      @output_buffer.push
+      result = render_template_for(@__vc_variant).to_s + _output_postamble
+      @output_buffer = orig_buf
+      @output_buffer.pop
+      result
+    end
+
+    def output_buffer=(other_buffer)
+      @output_buffer.replace(other_buffer)
+    end
+
+    def with_output_buffer(buf = nil)
+      unless buf
+        buf = ActionView::OutputBuffer.new
+        if output_buffer && output_buffer.respond_to?(:encoding)
+          buf.force_encoding(output_buffer.encoding)
+        end
+      end
+
+      output_buffer.push(buf)
+      result = nil
+
+      begin
+        yield
+      ensure
+        # assign result here to avoid a return statement, which will
+        # immediately return to the caller and swallow any errors
+        result = output_buffer.pop
+      end
+
+      result
+    end
+
+    module ActionViewMods
+      def with_output_buffer(buf = nil)
+        # save a reference to the stack-based buffer
+        orig_buf = output_buffer
+        result = nil
+
+        super do
+          if orig_buf.respond_to?(:push)
```

Comment:
Can we measure the performance difference between this and the original `with_output_buffer` implementation? My only concern is making existing rendering paths slower (specifically for dotcom), but it's possible it's comparable.

Reply:
It looks like things are slower with the patch, but the impact is negligible:

```
Warming up --------------------------------------
       without_patch     1.112k i/100ms
Calculating -------------------------------------
       without_patch     10.864k (± 4.6%) i/s -    108.976k in  10.055584s

Pausing here -- run Ruby again to measure the next benchmark...

Warming up --------------------------------------
          with_patch     1.077k i/100ms
Calculating -------------------------------------
          with_patch     10.629k (± 6.4%) i/s -    106.623k in  10.085791s

Comparison:
       without_patch:    10864.5 i/s
          with_patch:    10629.1 i/s - same-ish: difference falls within error
```